src/plugins/dashboard/public/attribute_service/attribute_service.tsx

```diff
@@ -0,0 +1,137 @@
+} from '../../../saved_objects/public';
+
+/**
+ * The attribute service is a shared, generic service that embeddables can use to provide the functionality
+ * required to fulfill the requirements of the ReferenceOrValueEmbeddable interface. The attribute_service
+ * can also be used as a higher level wrapper to transform an embeddable input shape that references a saved object
+ * into an embeddable input shape that contains that saved object's attributes by value.
+ */
+export class AttributeService<
+  SavedObjectAttributes,
+  ValType extends EmbeddableInput & { attributes: SavedObjectAttributes },
+  RefType extends SavedObjectEmbeddableInput
+> {
+  constructor(
+    private type: string,
+    private savedObjectsClient: SavedObjectsClientContract,
+    private i18nContext: I18nStart['Context']
+  ) {}
+
+  public async unwrapAttributes(input: RefType | ValType): Promise<SavedObjectAttributes> {
+    if (this.inputIsRefType(input)) {
+      const savedObject: SimpleSavedObject<SavedObjectAttributes> = await this.savedObjectsClient.get<
+        SavedObjectAttributes
+      >(this.type, input.savedObjectId);
+      return savedObject.attributes;
+    }
+    return input.attributes;
+  }
+
+  public async wrapAttributes(
+    newAttributes: SavedObjectAttributes,
+    useRefType: boolean,
+    embeddable?: IEmbeddable
+  ): Promise<Omit<ValType | RefType, 'id'>> {
+    const savedObjectId =
+      embeddable && isSavedObjectEmbeddableInput(embeddable.getInput())
+        ? (embeddable.getInput() as SavedObjectEmbeddableInput).savedObjectId
+        : undefined;
+
+    if (useRefType) {
+      if (savedObjectId) {
+        await this.savedObjectsClient.update(this.type, savedObjectId, newAttributes);
+        return { savedObjectId } as RefType;
+      } else {
+        const savedItem = await this.savedObjectsClient.create(this.type, newAttributes);
+        return { savedObjectId: savedItem.id } as RefType;
+      }
+    } else {
+      return { attributes: newAttributes } as ValType;
+    }
+  }
+
+  inputIsRefType = (input: ValType | RefType): input is RefType => {
+    return isSavedObjectEmbeddableInput(input);
+  };
+
+  getInputAsValueType = async (input: ValType | RefType): Promise<ValType> => {
+    if (!this.inputIsRefType(input)) {
+      return input;
+    }
+    const attributes = await this.unwrapAttributes(input);
+    return {
+      ...input,
+      savedObjectId: undefined,
+      attributes,
+    };
+  };
+
+  getInputAsRefType = async (
+    input: ValType | RefType,
+    saveOptions?: { showSaveModal: boolean } | { title: string }
+  ): Promise<RefType> => {
+    if (this.inputIsRefType(input)) {
+      return input;
+    }
+
+    return new Promise<RefType>((resolve, reject) => {
+      const onSave = async (props: OnSaveProps): Promise<SaveResult> => {
+        try {
+          const wrappedInput = (await this.wrapAttributes(input.attributes, true)) as RefType;
+          wrappedInput.title = props.newTitle;
```

Comment:
why is this assignment needed?

Reply:
The title is assigned here because adding an embeddable to the library via the attribute service shows a save modal with the opportunity to name (or re-name) the saved object.

Reply:
It is not a saved object rename, but just embeddable panel title change? isn't it? 
Is this expected?

Reply:
This modal shows up when adding an item to the library. That item (when it is by value) usually doesn't yet have a saved object title. 

When the item already has a saved object title, like in a case where it was previously in the library, you are given the option to change the name. An improvement here would be duplicate detection, which will be added in the follow up PR which adds the actual 'add to library' action.

For book embeddables, the titling is a little strange, which is what you're picking up on I believe. They have a saved Object title `input.title` and a book title `input.attributes.title`. In the add flyout, they are referred to by their book title.